p2p/state.py

```diff
@@ -161,32 +186,36 @@ def msg_queue_maxsize(self) -> int:
         # run in the background notice the cancel token has been triggered and return.
         await asyncio.sleep(0)
 
-    async def request_nodes(self, node_keys: List[bytes]) -> None:
-        batches = list(partition_all(eth.MAX_STATE_FETCH, node_keys))
-        for batch in batches:
-            peer = await self.get_idle_peer()
-            now = time.time()
-            for node_key in batch:
-                self._pending_nodes[node_key] = now
+    async def request_nodes(self, node_keys: Iterable[Hash32]) -> None:
+        not_yet_requested = set(node_keys)
+        while not_yet_requested:
+            try:
+                peer = await self.get_peer_for_request(not_yet_requested)
+            except NoEligiblePeers:
+                self.logger.debug(
+                    "No idle peers have any of the trie nodes we want, sleeping a bit")
+                await self.wait(asyncio.sleep(0.2))
+                continue
```

Comment:
If none of the peers we are connected to have a node we're looking for this loop will never terminate.

Reply:
Right... Maybe for now we could have another exception to signal that none of our peers have any of the nodes we're about to request. Then we could do one/more of the following:
1. If PeerPool is full, disconnect from one or more peers to make room for others that might have the nodes
2. Increase sleep delay
3. Request another batch of nodes

What do you think? Any other ideas?

Reply:
I'm fine punting on the long term solution for this (but lets document it with another issue).  I think option 3 is our best and simplest solution for now.

Reply:
Turns out it would take a big redesign of the whole `StateDownloader` in order to request a different batch as we don't have something like a back-burner area to put those node keys that no peers seem to have. I guess we'll have to go with option 1 and 2 when none our peers have the nodes. I'll give it some more thought, though, maybe I'll come up with something

Reply:
I went ahead and implemented the above so now when no peers have the nodes in a batch, they're added to a list of missing node keys and retried after a while.